ksql-server zombies if the rest api port is in use.

Taken from this thread: https://groups.google.com/forum/#!msg/confluent-platform/31ifMnUWEI4/C6gplgDWBAAJ

If you start up ksql-server when the rest api's port is already in use, then the app reports the exception:

```
[2018-04-19 11:30:15,953] WARN FAILED NetworkTrafficServerConnector@17f9344b{HTTP/1.1}{localhost:8088}: java.net.BindException: Address already in use (org.eclipse.jetty.util.component.AbstractLifeCycle:212)
java.net.BindException: Address already in use
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:433)
	at sun.nio.ch.Net.bind(Net.java:425)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
	at org.eclipse.jetty.server.ServerConnector.open(ServerConnector.java:321)
	at org.eclipse.jetty.server.AbstractNetworkConnector.doStart(AbstractNetworkConnector.java:80)
	at org.eclipse.jetty.server.ServerConnector.doStart(ServerConnector.java:236)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.server.Server.doStart(Server.java:366)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at io.confluent.rest.Application.start(Application.java:551)
	at io.confluent.ksql.rest.server.KsqlRestApplication.start(KsqlRestApplication.java:198)
	at io.confluent.ksql.rest.server.KsqlServerMain.main(KsqlServerMain.java:42)
[2018-04-19 11:30:15,954] WARN FAILED io.confluent.rest.Application$1@7bd69e82: java.net.BindException: Address already in use (org.eclipse.jetty.util.component.AbstractLifeCycle:212)
java.net.BindException: Address already in use
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:433)
	at sun.nio.ch.Net.bind(Net.java:425)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
	at org.eclipse.jetty.server.ServerConnector.open(ServerConnector.java:321)
	at org.eclipse.jetty.server.AbstractNetworkConnector.doStart(AbstractNetworkConnector.java:80)
	at org.eclipse.jetty.server.ServerConnector.doStart(ServerConnector.java:236)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.server.Server.doStart(Server.java:366)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at io.confluent.rest.Application.start(Application.java:551)
	at io.confluent.ksql.rest.server.KsqlRestApplication.start(KsqlRestApplication.java:198)
	at io.confluent.ksql.rest.server.KsqlServerMain.main(KsqlServerMain.java:42)
Exception in thread "main" java.net.BindException: Address already in use
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:433)
	at sun.nio.ch.Net.bind(Net.java:425)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
	at org.eclipse.jetty.server.ServerConnector.open(ServerConnector.java:321)
	at org.eclipse.jetty.server.AbstractNetworkConnector.doStart(AbstractNetworkConnector.java:80)
	at org.eclipse.jetty.server.ServerConnector.doStart(ServerConnector.java:236)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.server.Server.doStart(Server.java:366)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at io.confluent.rest.Application.start(Application.java:551)
	at io.confluent.ksql.rest.server.KsqlRestApplication.start(KsqlRestApplication.java:198)
	at io.confluent.ksql.rest.server.KsqlServerMain.main(KsqlServerMain.java:42)
```

But it doesn't shutdown. Instead is sits in an invalid state, making it hard for admins or monitoring tools to spot there is an issue.

This manifests to users when starting the CLI, where they will see the following:

```
**************** WARNING ******************
Remote server address may not be valid:
Error issuing GET to KSQL server

Caused by: java.net.SocketException: Connection reset
Caused by: Connection reset
*******************************************
```

--- a/ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlServerMain.java
+++ b/ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlServerMain.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright 2018 Confluent Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -20,6 +20,7 @@ import org.apache.kafka.streams.StreamsConfig;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Properties;
 import io.confluent.ksql.version.metrics.KsqlVersionCheckerAgent;
@@ -29,20 +30,39 @@ public class KsqlServerMain {
   private static final Logger log = LoggerFactory.getLogger(KsqlServerMain.class);
   private static final String KSQL_REST_SERVER_DEFAULT_APP_ID = "KSQL_REST_SERVER_DEFAULT_APP_ID";
 
-  public static void main(final String[] args) throws Exception {
-    final ServerOptions serverOptions = ServerOptions.parse(args);
-    if (serverOptions == null) {
-      return;
-    }
+  private final Executable executable;
 
-    final Properties properties = serverOptions.loadProperties(System::getProperties);
-    final Optional<String> queriesFile = serverOptions.getQueriesFile(properties);
-    final Executable executable = createExecutable(properties, queriesFile);
-    log.info("Starting server");
-    executable.start();
-    log.info("Server up and running");
-    executable.join();
-    log.info("Server shutting down");
+  public static void main(final String[] args) {
+    try {
+      final ServerOptions serverOptions = ServerOptions.parse(args);
+      if (serverOptions == null) {
+        return;
+      }
+
+      final Properties properties = serverOptions.loadProperties(System::getProperties);
+      final Optional<String> queriesFile = serverOptions.getQueriesFile(properties);
+      final Executable executable = createExecutable(properties, queriesFile);
+      new KsqlServerMain(executable).tryStartApp();
+    } catch (final Exception e) {
+      log.error("Failed to start KSQL", e);
+      System.exit(-1);
+    }
+  }
+
+  KsqlServerMain(final Executable executable) {
+    this.executable = Objects.requireNonNull(executable, "executable");
+  }
+
+  void tryStartApp() throws Exception {
+    try {
+      log.info("Starting server");
+      executable.start();
+      log.info("Server up and running");
+      executable.join();
+    } finally {
+      log.info("Server shutting down");
+      executable.stop();
+    }
   }
 
   @SuppressWarnings("OptionalUsedAsFieldOrParameterType")

--- a/ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/KsqlServerMainTest.java
+++ b/ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/KsqlServerMainTest.java
@@ -1,0 +1,82 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server;
+
+import org.easymock.EasyMockRunner;
+import org.easymock.Mock;
+import org.easymock.MockType;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import static org.easymock.EasyMock.expectLastCall;
+import static org.easymock.EasyMock.replay;
+import static org.easymock.EasyMock.verify;
+import static org.junit.Assert.*;
+
+/**
+ * @author andy
+ * created 19/04/2018
+ */
+@RunWith(EasyMockRunner.class)
+public class KsqlServerMainTest {
+  private KsqlServerMain main;
+
+  @Mock(MockType.NICE)
+  private Executable executable;
+
+  @Before
+  public void setUp() {
+    main = new KsqlServerMain(executable);
+  }
+
+  @Test
+  public void shouldStopAppOnJoin() throws Exception {
+    // Given:
+    executable.stop();
+    expectLastCall();
+    replay(executable);
+
+    // When:
+    main.tryStartApp();
+
+    // Then:
+    verify(executable);
+  }
+
+  @Test
+  public void shouldStopAppOnErrorStarting() throws Exception {
+    // Given:
+    executable.start();
+    expectLastCall().andThrow(new RuntimeException("Boom"));
+
+    executable.stop();
+    expectLastCall();
+    replay(executable);
+
+    try {
+      // When:
+      main.tryStartApp();
+      fail();
+    } catch (final Exception e) {
+      // Expected
+    }
+
+    // Then:
+    verify(executable);
+  }
+}

